### PR TITLE
Sound commands

### PR DESCRIPTION
**Pull request by Leinnan**
# Objective

Inspired a bit by the famous article about leaving Rust, I've created commands that simplify playing sounds in Bevy.

## Solution

- Create command for playing sounds

---

## Changelog

- Added new commands for playing sounds using existing API.

### CODE REVIEW

File: crates/bevy_audio/src/lib.rs [hunk 1]
```diff
@@ -119,3 +120,47 @@ impl AddAudioSource for App {
...
+        world.spawn(AudioBundle {
+            source,
+            settings: self.settings,
+        });
+    }
+}
+
+/// Trait for playing sounds with commands
+pub trait AudioSpawnCommandExt {
+    /// Command for playing a standard bevy audio asset with default settings
```
Comment (hymm) [inline crates/bevy_audio/src/lib.rs:146]: There should be a comment here that you need to preload the asset if you want the sound to play without delay.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I like the notion of an abstract "play sound" command, but I think it needs to take handles, not asset paths.

Like WrongShoe points out, it's really easy to introduce nasty delay (and repeated overhead) in the sounds due to asset loading time.

At that point though, I'm not actually sure this is simpler :(

**Comment by SolarLiner** [comment]
I also agree with the comments; I'll also add that this kind of shortcut, while nice, is not something we provide for any other features of the engine, so it stands out as a little weird to me.

The pattern of `Bundle { component: override, ..default() }` is common throughout the engine and is expected of users to learn fairly quickly as a pattern they can use to do what they want. 3rd party crates also use this pattern as a way to provide default behavior. It's certainly a bit more verbose than your solution, but it's also explicit, which makes it easier to read and review, and that makes it more in line with Rust's code guidelines as well.

File: crates/bevy_audio/src/lib.rs (hunk 1 above)
Comment (Leinnan) [reply crates/bevy_audio/src/lib.rs]: Done

**Comment by Leinnan** [comment]
> Like WrongShoe points out, it's really easy to introduce nasty delay (and repeated overhead) in the sounds due to asset loading time. That's a particularly frustrating footgun to discover later in your development lifecycle.

@alice-i-cecile I've updated the comment with information about possible delays. I agree on the fact that if we introduce handles to it, it wouldn't be simpler. What is useful IMHO in this PR is the fact that you don't need to pass `AssetServer` as system function parameter to be able to call it. If we want to use handles instead of `AssetPath` then I agree that it will mostly increase code size without significant improvements to developer experience. 

> The pattern of `Bundle { component: override, ..default() }` is common throughout the engine and is expected of users to learn fairly quickly as a pattern they can use to do what they want. 3rd party crates also use this pattern as a way to provide default behavior. It's certainly a bit more verbose than your solution, but it's also explicit, which makes it easier to read and review, and that makes it more in line with Rust's code guidelines as well.

In my mind, this PR was meant as an option to provide easier API for common cases and if it makes sense could be adapted to other places as well.

**Comment by alice-i-cecile** [comment]
The biggest concern I have here is that this is *always* wrong in a shipped game. The delays and constant reloading are real footguns, and we shouldn't introduce an API that seems appealingly simple but just doesn't work.

I definitely share your desire for a "please just play sound" API‚ but the existing design isn't actually coherent with how asset loading in Bevy works. In a world where we loaded everything from the assets folder immediately on project startup and never dropped anything, this would be perfect! And maybe there's an argument to be made for an easy asset-handling mode for jam-style games that supports this, but it's the wrong model for larger projects.

**Comment by BD103** [comment]
> The biggest concern I have here is that this is _always_ wrong in a shipped game. The delays and constant reloading are real footguns, and we shouldn't introduce an API that seems appealingly simple but just doesn't work.

I agree. I think that while this PR may quiet certain complaints, it does not solve the underlying issue. We need something like gizmos for one-shot audio, but I don't think `Commands` are the right way to solve it.

I appreciate the thought, but this is better put in a 3rd-party crate than inside of the main engine itself.

**Comment by Leinnan** [comment]
@alice-i-cecile @BD103 Since at the moment it could be easily be provided in separate package I am okay with closing the PR. Maybe I will create `crate` out of this and let the idea grow and mature. Later on, then it could become something that would be more aligned with the engine and become part of it. Or not, it is nothing wrong with that. ;)

**Pull request closed without merging**